File: trunk/Rationals/src/pt/iscte/dcti/pa/math/Rational.java
package pt.iscte.dcti.pa.math;

// TODO Extend class Number and implement conversion methods.

/*
 * Operations to implements:
 * Addition
 * Subtraction
 * Product
 * Division
 * Symmetrical
 * Inverse
 * equals
 * toString
 * compareTo
 */
public class Rational {

	private int numerator;
	private int denominator;

	public Rational(final int numerator, final int denominator) {
		if (denominator == 0)
			throw new IllegalArgumentException(
					"Illegal value of denominator. Should be != 0, was "
							+ denominator + ".");

		this.numerator = numerator;
		this.denominator = denominator;

		normalize();

		checkInvariant();
	}

	public Rational(final int number) {
		this(number, 1);
	}

	public Rational() {
		this(0, 1);
	}

	public int getNumerator() {
		checkInvariant();

		return numerator;
	}

	public int getDenominator() {
		checkInvariant();

		return denominator;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + denominator;
		result = prime * result + numerator;
		return result;
	}

	@Override
	public boolean equals(Object object) {
		if (this == object)
			return true;
		if (object == null)
			return false;
		if (getClass() != object.getClass())
			return false;
		
		Rational other = (Rational) object;
		
		return denominator == other.denominator && numerator == other.numerator;
	}

	public void add(Rational r1) {
		// TODO Auto-generated method stub
		
	}

	private void checkInvariant() {
		assert 0 < denominator : "Illegal state. Denominator should be > 0, is "
				+ denominator + ".";
		assert gcd(numerator, denominator) == 1 : "Illegal state. gcd(numerator, denominator) should be 1, is "
				+ gcd(numerator, denominator) + ".";
	}

	private void normalize() {
		if (numerator == 0 && denominator == 0)
			return;

		if (denominator < 0) {
			numerator = -numerator;
			denominator = -denominator;
		}

		int gcd = gcd(numerator, denominator);

		numerator /= gcd;
		denominator /= gcd;
	}

	private static int gcd(int firstValue, int secondValue) {
		assert firstValue != 0 || secondValue != 0 : "IIlegal value of arguments. Cannot be both zero.";

		if (firstValue < 0)
			firstValue = -firstValue;
		if (secondValue < 0)
			secondValue = -secondValue;

		while (firstValue != 0) {
			final int oldFirstValue = firstValue;
			firstValue = secondValue % firstValue;
			secondValue = oldFirstValue;
		}

		return secondValue;
	}

	// TODO Remove main from Rational class.
	public static void main(final String[] arguments) {
		new Rational(2, 4);
	}

}

File: trunk/Rationals (C++)/Rational.cpp
#include "Rational.hpp"

pt::iscte::dcti::pa::Rational::Rational(int const numerator, int const denominator) :
	numerator_(numerator), denominator_(denominator) {
	assert(denominator != 0);

	if (denominator < 0) {
		numerator_ = -numerator;
		denominator_ = -denominator;
	}

	reduce();

	checkInvariant();
}

int pt::iscte::dcti::pa::Rational::gcd(int firstValue, int secondValue) {
	assert(firstValue != 0 || secondValue != 0);

	if (firstValue < 0)
		firstValue = -firstValue;
	if (secondValue < 0)
		secondValue = -secondValue;

	while (firstValue != 0) {
		int const oldFirstValue = firstValue;
		firstValue = secondValue % firstValue;
		secondValue = oldFirstValue;
	}

	return secondValue;
}

void pt::iscte::dcti::pa::Rational::operator+=(Rational otherRational) {
	checkInvariant();

	numerator_ = otherRational.denominator() * numerator() + denominator() * otherRational.numerator();
	denominator_ *= otherRational.denominator();

	reduce();

	checkInvariant();
}

bool pt::iscte::dcti::pa::Rational::operator==(Rational otherRational) const {
	checkInvariant();

	return numerator() == otherRational.numerator() && denominator() == otherRational.denominator();
}


#include <iostream>

using namespace std;
using namespace pt::iscte::dcti::pa;

int main() {
	cout << "Starting tests." << endl;

	cout << "Constructing r1." << endl;
	Rational r1(3, 2);

	cout << "Constructing r2." << endl;
	Rational r2(1, 2);

	cout << "Adding r2 to r1." << endl;
	r1 += r2;

	cout << "Assigning r1 to r2." << endl;
	r2 = r1;

	cout << "Checking correction of addition." << endl;
	assert(r1 == 2);

	Rational const r3(1, 3);

	cout << r3.numerator() << endl;
}


